### PR TITLE
virt-api: Apply defaults to VM returned by expand-spec

### LINKED ISSUE #12360: expand-spec subresource API output missing default devices
/area instancetype

**What happened**:

$subject, the current behaviour of the expand-spec subresource API is just to expand any referenced instance type and preference. This isn't useful for end users and we have now had requests to include the default devices in the output from consumers of the API downstream.

**What you expected to happen**:

Default devices to be included in the output of the API with any referenced preferences applied.

**How to reproduce it (as minimally and precisely as possible)**:

Call API with VM definition or against a VM referencing an instance type and preference.

**Additional context**:

https://issues.redhat.com/browse/CNV-41608

**Environment**:
- KubeVirt version (use `virtctl version`): N/A
- Kubernetes version (use `kubectl version`): N/A
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by lyarwood**
/area instancetype

### What this PR does
Before this PR:

Previously the VM returned by the expand-spec subresource API would only expand any referenced instance types and preferences. 

```
$ ./cluster-up/kubectl.sh apply -k https://github.com/kubevirt/common-instancetypes.git
[..]
$ ./cluster-up/virtctl.sh create vm --instancetype u1.medium --volume-containerdisk src:quay.io/containerdisks/fedora:40 > fedora.yaml
$ ./cluster-up/virtctl.sh expand --file fedora.yaml 
selecting podman as container runtime
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  creationTimestamp: null
  name: vm-n9hks
  namespace: default
spec:
  runStrategy: Always
  template:
    metadata:
      creationTimestamp: null
    spec:
      domain:
        cpu:
          cores: 1
          sockets: 1
          threads: 1
        devices: {}
        memory:
          guest: 4Gi
        resources: {}
      terminationGracePeriodSeconds: 180
      volumes:
      - containerDisk:
          image: quay.io/containerdisks/fedora:40
        name: vm-n9hks-containerdisk-0
status: {}
```

After this PR:

This change now applies defaults to the VM and more importantly the VMI spec within to provide a more complete representation of the eventual VMI created from the VM.

```
./cluster-up/virtctl.sh expand --file fedora.yaml 
selecting podman as container runtime
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  creationTimestamp: null
  name: vm-n9hks
  namespace: default
spec:
  runStrategy: Always
  template:
    metadata:
      creationTimestamp: null
    spec:
      architecture: amd64
      domain:
        cpu:
          cores: 1
          model: host-model
          sockets: 1
          threads: 1
        devices:
          disks:
          - disk:
              bus: sata
            name: vm-n9hks-containerdisk-0
          interfaces:
          - bridge: {}
            name: default
        machine:
          type: q35
        memory:
          guest: 4Gi
        resources:
          requests:
            memory: 4Gi
      evictionStrategy: None
      networks:
      - name: default
        pod: {}
      terminationGracePeriodSeconds: 180
      volumes:
      - containerDisk:
          image: quay.io/containerdisks/fedora:40
          imagePullPolicy: IfNotPresent
        name: vm-n9hks-containerdisk-0
status: {}
```

Fixes #12360

### Why we need it and why it was done in this way

The following tradeoffs were made:

There are numerous cleanups that I've deferred with this change as I'd like to backport this change in functionality as a bugfix to at least `release-1.1`. I'll address these and some cleanups in a follow up series.

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
The `expand-spec` subresource API now applies defaults to the returned `VirtualMachine` to ensure the `VirtualMachineInstanceSpec` within is closer to the eventual version used when starting the original `VirtualMachine`.
```

### CODE REVIEW

**Review by 0xFelix — APPROVED** [review]
/lgtm

File: pkg/virt-api/rest/expand.go [hunk 1]
```diff
@@ -101,12 +103,25 @@ func expandSpecResponse(vm *v1.VirtualMachine, instancetypeMethods instancetype.
 		return
 	}
 
-	conflicts := instancetypeMethods.ApplyToVmi(field.NewPath("spec", "template", "spec"), instancetypeSpec, preferenceSpec, &vm.Spec.Template.Spec, &vm.Spec.Template.ObjectMeta)
+	util.SetDefaultVolumeDisk(&vm.Spec.Template.Spec)
+
+	if err := vmispec.SetDefaultNetworkInterface(app.clusterConfig, &vm.Spec.Template.Spec); err != nil {
+		writeError(errorFunc(err), response)
+		return
+	}
```
Comment (xpivarc) [inline pkg/virt-api/rest/expand.go:111]: Q: Could we reuse dry-run create?

Comment (lyarwood) [reply pkg/virt-api/rest/expand.go:111]: @xpivarc  Unfortunately not, we only apply instance types and preferences through the VirtualMachine controller. Even reusing dry-run create for these defaults wouldn't work as others would also be set (vCPU topology, resources etc) that would conflict with any provided the instance type. 

So unfortunately we end up duplicating most of the logic from the controller here. I do want to clean this up in a follow up series once this change has landed and been backported.

Comment (xpivarc) [reply pkg/virt-api/rest/expand.go:111]: >So unfortunately we end up duplicating most of the logic from the controller here. I do want to clean this up in a follow up series once this change has landed and been backported.

Looking forward

**Review by xpivarc — COMMENTED** [review]
/approve

**Pull request merged**